### PR TITLE
doc: rewrite dependencies.md

### PR DESCRIPTION
**Pull request by fanquake**
This PR rewrites dependencies.md. The current list is hard to parse, includes information that is either incorrect and/or misleading, and duplicates info in other documentation. The list of dependencies is much smaller, because it's now just the actual dependencies of Bitcoin Core, not random Qt things, or the dependencies of other tooling. We don't need _another_ section on configure flag usage, or, to have duplicated lists of dependencies in other build docs, as that somewhat defeats the point of having dependencies.md, and just means more effort keeping things in sync.

### CODE REVIEW

File: doc/dependencies.md [hunk 1]
```diff
@@ -1,51 +1,53 @@
...
+
+## Build, Debug & Tests
+
+A compiler with support for C++17 and `std::filesystem` is required.
+
+When compiling against `-stdlib=libc++`, version 7.0 or later is required.
+
+| Compiler | Minimum required |
+| --- | --- |
+| [Clang](https://clang.llvm.org) | [7.0](https://github.com/bitcoin/bitcoin/pull/23060) |
```
Comment (katesalazar) [inline doc/dependencies.md:15]: I'd love tables free from links:
```
| Compiler   | Minimum required |
| ---------- | ---------------- |
| [Clang][0] | [7.0][1]         |
```

```
[0]: https://clang.llvm.org
[1]: https://github.com/bitcoin/bitcoin/pull/23060
```

Much better than:
```
| Compiler | Minimum required |
| --- | --- |
| [Clang](https://clang.llvm.org) | [7.0](https://github.com/bitcoin/bitcoin/pull/23060) |
```

**Comment by mjdietzx** [comment]
Concept ACK - nice cleanup

**Review by jarolrod — COMMENTED** [review]
Concept ACK

> to have duplicated lists of dependencies in other build docs

I had originally duplicated the lists in the other build docs, besides build-unix.md. Since we have dependencies.md, I'd be in agreement of keeping dependencies there. My idea with the dependencies list in the build docs was to keep most of the information that a user needs to build contained within the build doc itself, and not need to jump around. I understand that this approach can be an unnecessary maintenance cost, especially when these dependencies change.

Do you want to go ahead and remove the dependency tables from the other build-*.md docs?

**Comment by fanquake** [comment]
> Do you want to go ahead and remove the dependency tables from the other build-*.md docs?

Have done.

**Comment by naumenkogs** [comment]
Concept ACK

File: doc/dependencies.md [hunk 2]
```diff
@@ -1,51 +1,53 @@
...
+| [Boost](https://www.boost.org/users/download/) | 1.71.0 | [1.64.0](https://github.com/bitcoin/bitcoin/pull/22320) | No |
+| [libevent](https://github.com/libevent/libevent/releases) | 2.1.12-stable | [2.0.21](https://github.com/bitcoin/bitcoin/pull/18676) | No |
+| [glibc](https://www.gnu.org/software/libc/) | N/A | [2.18](https://github.com/bitcoin/bitcoin/pull/23511) | Yes |
+
+## Optional
+
+### GUI
+| Dependency | Version used | Minimum required | Runtime |
+| --- | --- | --- | --- |
+| [Fontconfig](https://www.freedesktop.org/wiki/Software/fontconfig/) | 2.12.1 | 2.6 | Yes |
```
Comment (shaavan) [inline doc/dependencies.md:33]: **nit:** Since the package name is **fontconfig**. I would suggest updating the name to:
```suggestion
| [fontconfig](https://www.freedesktop.org/wiki/Software/fontconfig/) | 2.12.1 | 2.6 | Yes |
```

**Review by shaavan — COMMENTED** [review]
Concept ACK

- Cleanup looks good. The list displays all the significant dependencies for compiling **Bitcoin Core**. 
- I have checked that the information provided in update `dependencies.md` is unchanged from the master’s `dependencies.md`.
- **MiniUPnPc** minimum version required info is added. Unfortunately, I cannot test whether this information is correct as I am on **Ubuntu 20.04**, and the package version below _2.1.20_ is not available on it.
- The links are correct and direct to the right appropriate web pages.

File: doc/dependencies.md (hunk 1 above)
Comment (jamesob) [reply doc/dependencies.md:15]: Not necessary IMO, and matching up the indexes may get confusing if there is a long list of footnotes. Might be better  if the indexes could be strings and not numbers, but still unnecessary IMO.

File: doc/build-osx.md [hunk 3]
```diff
@@ -6,41 +6,6 @@ This guide describes how to build bitcoind, command-line utilities, and GUI on m
...
-The following dependencies are **optional**:
-
-Library                                                         | Purpose          | Description
---------------------------------------------------------------- |------------------|----------------------
-[berkeley-db@4](https://formulae.brew.sh/formula/berkeley-db@4) | Berkeley DB      | Wallet storage (only needed when wallet enabled)
-[qt@5](https://formulae.brew.sh/formula/qt@5)                   | GUI              | GUI toolkit (only needed when GUI enabled)
-[qrencode](https://formulae.brew.sh/formula/qrencode)           | QR codes in GUI  | Generating QR codes (only needed when GUI enabled)
-[zeromq](https://formulae.brew.sh/formula/zeromq)               | ZMQ notification | Allows generating ZMQ notifications (requires ZMQ version >= 4.0.0)
-[sqlite](https://formulae.brew.sh/formula/sqlite)               | SQLite DB        | Wallet storage (only needed when wallet enabled)
-[miniupnpc](https://formulae.brew.sh/formula/miniupnpc)         | UPnP Support     | Firewall-jumping support (needed for port mapping support)
```
Comment (jamesob) [inline doc/build-osx.md:30]: So far as I can tell, for deps like `miniupnpc` we no longer have the "Description" information living here. Might be nice to add a column to the dependencies.md table that lists the info here.

**Review by jamesob — COMMENTED** [review]
Concept ACK

File: doc/dependencies.md (hunk 2 above)
Comment (katesalazar) [reply doc/dependencies.md:33]: Even its own web site uses capitalized and non capitalized randomly. Use whatever fits other rows.

**Review by RandyMcMillan — COMMENTED** [review]
https://github.com/bitcoin/bitcoin/pull/23565/files/e57aca9359bdbb8813c0bea9df92e81b97dfe00d#diff-e9ba5bd8b63e72aa2d54d02960c9209823938981691cc8786ba323b54de0659cR53

`bison` needs to be installed for the /depends build. (yacc)
Maybe include it here for completeness?

**Comment by jarolrod** [comment]
@RandyMcMillan the mentioned build docs and this PR are for dynamic builds, not static builds. The bison dependency for static builds is already documented under the /depends readme.

File: doc/build-osx.md [hunk 4]
```diff
@@ -79,6 +44,9 @@ Note: If you run into issues while installing Homebrew or pulling packages, refe
 
 The first step is to download the required dependencies.
 These dependencies represent the packages required to get a barebones installation up and running.
+
+See [dependencies.md](dependencies.md) for a complete overview.
+
 To install, run the following from your terminal:
 
 ``` bash
```
Comment (RandyMcMillan) [inline doc/build-osx.md:52]: `bison` needs to be installed for the /depends build. (yacc)
Maybe include it here for completeness?

File: doc/build-freebsd.md [hunk 5]
```diff
@@ -4,35 +4,6 @@
 
 This guide describes how to build bitcoind, command-line utilities, and GUI on FreeBSD.
 
-## Dependencies
-
-The following dependencies are **required**:
-
- Library                                                               | Purpose    | Description
- ----------------------------------------------------------------------|------------|----------------------
- [autoconf](https://svnweb.freebsd.org/ports/head/devel/autoconf/)     | Build      | Automatically configure software source code
```
[the remark this answers is not in the document]
Comment (jarolrod) [reply doc/build-freebsd.md:13]: No, it is not drawn in by another dependency. This is a situation that highlights that no approach is perfect.

The current duplicated tables have certain specific dependencies that only apply to that platform. With the approach introduced in this PR, you'd have to list the dependency and note that it applies for X platform.

**Comment by RandyMcMillan** [comment]
@jarolrod - I get it - but to reduce documentation complexity further - we could download everything for a static build - then it is just a matter of enumerating the different commands for dynamic or static builds. Each platform would need one build doc.

File: doc/build-freebsd.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply doc/build-freebsd.md:13]: > Looks like none of the autotools stuff is mentioned in the new deps section 

It isn't, and I don't think it needs to be. It's not a dependency of Bitcoin Core, just like `git`, or `hexdump` isn't. They are build tools, and the installation instructions for them are is the build docs.

> The current duplicated tables have certain specific dependencies that only apply to that platform.

That's a rarity, and I think it's fine for those to just be mentioned in the build docs where relevant.

[the remark this answers is not in the document]
Comment (katesalazar) [reply doc/build-freebsd.md:13]: > > Looks like none of the autotools stuff is mentioned in the new deps section
> 
> It isn't, and I don't think it needs to be. It's not a dependency of Bitcoin Core, just like `git`, or `hexdump` isn't. They are build tools, and the installation instructions for them are is the build docs.

autotools are build tools;
when no alternative to them,
they are build dependencies

**Comment by Rspigler** [comment]
Concept ACK.  Cleanup is nice

File: doc/dependencies.md [hunk 6]
```diff
@@ -1,48 +1,55 @@
...
+| [Python](https://www.python.org) (tests) | [3.6](https://github.com/bitcoin/bitcoin/pull/19504) |
+| [systemtap](https://sourceware.org/systemtap/) ([tracing](tracing.md))| N/A |
+
+## Required
+
+| Dependency | Version used | Minimum required | Runtime |
+| --- | --- | --- | --- |
+| [Boost](https://www.boost.org/users/download/) | 1.71.0 | [1.64.0](https://github.com/bitcoin/bitcoin/pull/22320) | No |
+| [libevent](https://github.com/libevent/libevent/releases) | 2.1.12-stable | [2.0.21](https://github.com/bitcoin/bitcoin/pull/18676) | No |
+| [glibc](https://www.gnu.org/software/libc/) | N/A | [2.18](https://github.com/bitcoin/bitcoin/pull/23511) | Yes |
```
Comment (luke-jr) [inline doc/dependencies.md:28]: AFAIK 2.18 is only required for the release binaries, not the minimum required.

File: doc/dependencies.md [hunk 7]
```diff
@@ -1,48 +1,55 @@
...
+
+These are the dependencies used by Bitcoin Core.
+You can find installation instructions in the `build-*.md` file for your platform.
+"Runtime" and "Version Used" are both in reference to the release binaries.
+
+## Build, Debug & Tests
+
+A compiler with support for C++17 and `std::filesystem` is required.
+
+When compiling against `-stdlib=libc++`, version 7.0 or later is required.
```
Comment (MarcoFalke) [inline doc/dependencies.md:11]: Not sure if this is needed. I think this should only be mentioned when the required libc++ version is *larger* than the required clang version.
It should be self-explanatory that you can't use a libc++ version *smaller* than your clang version.

Comment (fanquake) [reply doc/dependencies.md]: Agree. Dropped this line.

File: doc/dependencies.md (hunk 6 above)
Comment (fanquake) [reply doc/dependencies.md:26]: It could potentially be `2.16`, if you build with `--disable-threadlocal`, however I think using `2.18`, which aligns with the release binairies, is fine here.

**Comment by fanquake** [comment]
Rebased.

**Comment by fanquake** [comment]
Rebased past #24132.

**Comment by fanquake** [comment]
Rebased past #24164.

File: doc/build-freebsd.md [hunk 8]
```diff
@@ -4,35 +4,6 @@
...
-
-
-The following dependencies are **optional**:
-
-  Library                                                                    | Purpose          | Description
-  ---------------------------------------------------------------------------|------------------|----------------------
-  [db5](https://svnweb.freebsd.org/ports/head/databases/db5/)                | Berkeley DB      | Wallet storage (only needed when wallet enabled)
-  [qt5](https://svnweb.freebsd.org/ports/head/devel/qt5/)                    | GUI              | GUI toolkit (only needed when GUI enabled)
-  [libqrencode](https://svnweb.freebsd.org/ports/head/graphics/libqrencode/) | QR codes in GUI  | Generating QR codes (only needed when GUI enabled)
-  [libzmq4](https://svnweb.freebsd.org/ports/head/net/libzmq4/)              | ZMQ notification | Allows generating ZMQ notifications (requires ZMQ version >= 4.0.0)
```
Comment (MarcoFalke) [inline doc/build-freebsd.md:30]: I think it is useful to know the name of the optional dependency. Maybe add a new line on how to install all the optional dependencies?

```
pkg install libzmq4 ...

Comment (fanquake) [reply doc/build-freebsd.md:30]: I've added a line for libzmq4, all of the others are already mentioned.

**Review by maflcko — APPROVED** [review]
cr ACK 893e18059f1f92a254aa2026312a769c0e711db6

Nice cleanup

**Review by hebasto — APPROVED** [review]
ACK 893e18059f1f92a254aa2026312a769c0e711db6, I have reviewed the code and it looks OK, I agree it can be merged.

I've verified links, package versions -- both in depends, and minimum required. GUI dependencies verified against Qt docs.

**Review by jarolrod — COMMENTED** [review]
crACK https://github.com/bitcoin/bitcoin/commit/893e18059f1f92a254aa2026312a769c0e711db6

**Comment by fanquake** [comment]
Thanks for the reviews. Planning on open some further doc consolidation PRs shortly.
